wheatley/row_generation/helpers.py

```diff
@@ -74,3 +74,21 @@ def convert_to_bell_string(bell: int) -> str:
 def rounds(number_of_bells: int) -> Row:
     """ Generate rounds on the given number of bells. """
     return Row([Bell.from_number(i) for i in range(1, number_of_bells + 1)])
+
+
+def generate_starting_row(number_of_bells: int, start_row: str = None) -> Row:
+    """ Generate the starting row as rounds or a custom input. """
+    if start_row is None:
+        return rounds(number_of_bells)
+
+    custom_start_row = [Bell.from_str(i) for i in start_row]
+    if len(custom_start_row) > len(set(custom_start_row)):
+        raise ValueError(f"starting row '{start_row}' contains the same bell multiple times")
+
+    # If there are more bells than given in the starting row
+    # add the missing ones in sequential order as cover bells
+    for i in range(1, number_of_bells + 1):
```

Comment:
Currently if you give this a bigger bell (e.g. `generate_starting_row(4, "T")` it generates `T1234` instead of `T1234567890E`

```py
self.assertListEqual(Row([Bell(5), Bell(0), Bell(1), Bell(2), Bell(3), Bell(4)]), generate_starting_row(4, "6"))
```

Needs something like
```py
    highest_given_bell = max(bell.number for bell in custom_start_row)
    upper_limit = max(number_of_bells, highest_given_bell) + 1
    for i in range(1, upper_limit):
```

Reply:
This one was on purpose, my thinking was that number_of_bells is the key thing, we should never have bells that are higher than that.

So there are 2 cases I think.
We start with some case (your example) `generate_starting_row(4, "6")` => `64321`

Then either:
* In bot it turns out there are actually 6 bells in the tower, this then gets rerun as `generate_starting_row(4, "6")` => `654321`
* There are actually only 4 (or any number fewer than 6) - then in _check_starting_row the validation `len(self._opening_rounds) != self.number_of_bells` will fail and tell the user that the starting row is too long.

I was thinking the most common case is just a mistake where you specify a number too high.
I grant the error is not the best, but it also catches where you just enter a string that's too long and I was not sure of an easy way to separate the validations

Reply:
An advantage of padding out to the max bell would be it always returns a valid row. Checking `len(self._opening_rounds) != self.number_of_bells` would still mean if the entered string contained `T` then the tower would need at least 12 bells.

There are 2 places this method is called, once in row_generator and again in the bot. The bot updates opening rounds as tower size changes, row_generator doesn't need to update once it is created. 

Currently the subtle bug I hit was is with ` -m "Plain Hunt on 4" --start-row 6` in a 6 bell tower
> INFO:BOT:ROW: 6 1 2 3 4 5
INFO:BOT:Starting to ring Plain Hunt on 4
INFO:BOT:ROW: 6 1 2 3 4 5
INFO:BOT:ROW: 1 6 3 2 4 6
ERROR:TOWER:Bell 6 on opposite stroke

If you add `--verbose` you get
> DEBUG:ROWGEN:Reset
DEBUG:ROWGEN:[1, 6, 3, 2, 4]
INFO:BOT:ROW: 1 6 3 2 4 6

row_generator calls `generate_starting_row(4, "6")` and gets back a 5 bell list `61234` for the starting row.
It uses this row and permutes it with the methods place notation up to its stage (i.e. it preserves cover bells it knows about)
As the 5 bell row returned is less than the 6 bells in the tower, cover bells are added from rounds
https://github.com/kneasle/wheatley/blob/5eb2447c1ea6fb170d001615939f98e36fd37aeb/wheatley/bot.py#L275-L278

- Line 278 should now add cover bells from `self.opening_rounds`

Reply:
OK right I haven't checked in on this for a while - I was working on #162 last night and didn't have time to look at this.  Generally, things seem to be going really well!

But I think that the weird bugs that are being discovered are more evidence that throwing a hard error while in the `main` function is the correct response to the user entering an invalid row (`143` is invalid because it doesn't contain `2` even if Wheatley tries to coerce the user's inputs to be valid).  I just think that most people wouldn't expect this behaviour, and ultimately I will go with whatever Wheatley's users think is most intuitive (also very few people actually read any documentation - I know this from how many people have asked questions which are answered in the docs) - to that end I'm tempted to make a poll on Wheatley Whisperers or the main ringing group to see what people think. @centreboard @chrisjfield what do you think to that idea?

It's also consistent with how all the other CLI args work (they all have quite strict parsing rules to reject ambiguous or incorrect inputs).  It also wouldn't be too hard to make this change (I think the resulting code might even be simpler).

Reply:
I think thats ok.

The only downside is if you want a non-standard covering bell you have to specify it manually - but not sure thats a huge issue really.

Reply:
Added in a parser and some tests on it, should now force you to enter a string containing all valid bells between 1 and the highest bell specified (i.e. the length of the string)

Reply:
Awesome!  I'll checkout this branch on local and have a play around.